common/docker/configuration.go

```diff
@@ -0,0 +1,335 @@
+}
+
+func getCardinalContainerName(cfg *config.Config) string {
+	return fmt.Sprintf("%s-cardinal", cfg.DockerEnv["CARDINAL_NAMESPACE"])
+}
+
+func getNakamaContainerName(cfg *config.Config) string {
+	return fmt.Sprintf("%s-nakama", cfg.DockerEnv["CARDINAL_NAMESPACE"])
+}
+
+func getNakamaDBContainerName(cfg *config.Config) string {
+	return fmt.Sprintf("%s-nakama-db", cfg.DockerEnv["CARDINAL_NAMESPACE"])
+}
+
+func getCelestiaDevNetContainerName(cfg *config.Config) string {
+	return fmt.Sprintf("%s-celestia-devnet", cfg.DockerEnv["CARDINAL_NAMESPACE"])
+}
+
+func getEVMContainerName(cfg *config.Config) string {
+	return fmt.Sprintf("%s-evm", cfg.DockerEnv["CARDINAL_NAMESPACE"])
+}
+
+func Redis(cfg *config.Config) Config {
+	redisPort := cfg.DockerEnv["REDIS_PORT"]
+	if redisPort == "" {
+		redisPort = "6379"
+	}
+
+	return Config{
+		Name: getRedisContainerName(cfg),
+		Config: &container.Config{
+			Image: "redis:latest",
+			Env: []string{
+				fmt.Sprintf("REDIS_PASSWORD=%s", cfg.DockerEnv["REDIS_PASSWORD"]),
+			},
+			ExposedPorts: nat.PortSet{
+				"6379/tcp": struct{}{},
+			},
+		},
+		HostConfig: &container.HostConfig{
+			PortBindings: nat.PortMap{
+				"6379/tcp": []nat.PortBinding{{HostPort: redisPort}},
+			},
+			RestartPolicy: container.RestartPolicy{Name: "unless-stopped"},
+			Mounts:        []mount.Mount{{Type: mount.TypeVolume, Source: "data", Target: "/redis"}},
+			NetworkMode:   container.NetworkMode(cfg.DockerEnv["CARDINAL_NAMESPACE"]),
+		},
+	}
+}
+
+func Cardinal(cfg *config.Config) Config {
+	imageName := cfg.DockerEnv["CARDINAL_NAMESPACE"]
+	containerConfig := container.Config{
+		Image: imageName,
+		Env: []string{
+			fmt.Sprintf("REDIS_ADDRESS=%s:6379", getRedisContainerName(cfg)),
+			fmt.Sprintf("BASE_SHARD_SEQUENCER_ADDRESS=%s:9601", getEVMContainerName(cfg)),
+		},
+		ExposedPorts: nat.PortSet{
+			"4040/tcp": struct{}{},
+		},
+	}
+
+	hostConfig := container.HostConfig{
+		PortBindings: nat.PortMap{
+			"4040/tcp": []nat.PortBinding{{HostPort: "4040"}},
+		},
+		RestartPolicy: container.RestartPolicy{Name: "unless-stopped"},
+		NetworkMode:   container.NetworkMode(cfg.DockerEnv["CARDINAL_NAMESPACE"]),
+	}
+
+	dockerfile := Dockerfile{
+		Script: cardinalDockerfile,
+		Target: "runtime",
+		// Images for the build
+		PrerequisiteImages: []Config{
+			{
+				Name: "golang:1.22-bookworm",
+				Config: &container.Config{
+					Image: "golang:1.22-bookworm",
+				},
+			},
+			{
+				Name: "gcr.io/distroless/base-debian12",
+				Config: &container.Config{
+					Image: "gcr.io/distroless/base-debian12",
+				},
+			},
+		},
+	}
+
+	// Add debug options
+	debug := cfg.Debug
+	if debug {
+		containerConfig.ExposedPorts["40000/tcp"] = struct{}{}
+		hostConfig.PortBindings["40000/tcp"] = []nat.PortBinding{{HostPort: "40000"}}
+		hostConfig.CapAdd = []string{"SYS_PTRACE"}
+		hostConfig.SecurityOpt = []string{"seccomp:unconfined"}
+		dockerfile.Target = "runtime-debug"
+	}
+
+	return Config{
```

Comment:
I don't have preference on whether to declare the Config, HostConfig, and Dockerfile directly inside Config or not, but I'd try to be consistent. Notice how you do it differently for Cardinal vs. Redis

Reply:
Moved the Cardinal declaration directly inside Config/Service to make it consistent (same with other service/config declaration)